File: 05-go-sse/kick.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

const (
	kickPusherURL = "wss://ws-us2.pusher.com/app/32cbd69e4b950bf97679?protocol=7&client=js&version=8.4.0&flash=false"
	browserUA     = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
)

// StartKick resolves chatroom ids for slugs (or uses overrides) and connects to
// Kick's Pusher socket, reconnecting with backoff. chatroomIDs aligns to slugs.
func StartKick(ctx context.Context, hub *Hub, slugs []string, chatroomIDs []string) {
	if len(slugs) == 0 {
		return
	}

	// Resolve chatroom ids -> slug map for channel labels.
	rooms := map[string]string{} // chatroomID -> slug label
	for i, slug := range slugs {
		slug = strings.TrimSpace(slug)
		if slug == "" {
			continue
		}
		var id string
		if i < len(chatroomIDs) && strings.TrimSpace(chatroomIDs[i]) != "" {
			id = strings.TrimSpace(chatroomIDs[i])
		} else {
			resolved, err := resolveKickChatroom(ctx, slug)
			if err != nil {
				log.Printf("[kick] could not resolve chatroom for %q: %v (set KICK_CHATROOM_IDS to override)", slug, err)
				continue
			}
			id = resolved
		}
		rooms[id] = slug
	}
	if len(rooms) == 0 {
		log.Printf("[kick] no chatrooms resolved; kick connector idle")
		return
	}

	go func() {
		backoff := time.Second
		for {
			if ctx.Err() != nil {
				return
			}
			err := kickSession(ctx, hub, rooms)
			if ctx.Err() != nil {
				return
			}
			if err != nil {
				log.Printf("[kick] session ended: %v (reconnect in %s)", err, backoff)
			}
			select {
			case <-ctx.Done():
				return
			case <-time.After(backoff):
			}
			backoff *= 2
			if backoff > 30*time.Second {
				backoff = 30 * time.Second
			}
		}
	}()
}

func resolveKickChatroom(ctx context.Context, slug string) (string, error) {
	url := "https://kick.com/api/v2/channels/" + slug
	reqCtx, cancel := context.WithTimeout(ctx, 12*time.Second)
	defer cancel()
	req, _ := http.NewRequestWithContext(reqCtx, http.MethodGet, url, nil)
	req.Header.Set("User-Agent", browserUA)
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("status %d", resp.StatusCode)
	}
	body, _ := io.ReadAll(resp.Body)
	var parsed struct {
		Chatroom struct {
			ID int64 `json:"id"`
		} `json:"chatroom"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return "", err
	}
	if parsed.Chatroom.ID == 0 {
		return "", fmt.Errorf("no chatroom id in response")
	}
	return fmt.Sprintf("%d", parsed.Chatroom.ID), nil
}

func kickSession(ctx context.Context, hub *Hub, rooms map[string]string) error {
	dialCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()
	conn, _, err := websocket.DefaultDialer.DialContext(dialCtx, kickPusherURL, nil)
	if err != nil {
		return fmt.Errorf("dial: %w", err)
	}
	defer conn.Close()

	go func() {
		<-ctx.Done()
		_ = conn.Close()
	}()

	writeJSON := func(v any) error { return conn.WriteJSON(v) }

	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			return fmt.Errorf("read: %w", err)
		}

		var frame struct {
			Event   string `json:"event"`
			Channel string `json:"channel"`
			Data    string `json:"data"`
		}
		if err := json.Unmarshal(data, &frame); err != nil {
			continue
		}

		switch {
		case frame.Event == "pusher:connection_established":
			for id := range rooms {
				sub := map[string]any{
					"event": "pusher:subscribe",
					"data": map[string]string{
						"auth":    "",
						"channel": "chatrooms." + id + ".v2",
					},
				}
				_ = writeJSON(sub)
			}
			log.Printf("[kick] connected, subscribed to %d chatroom(s)", len(rooms))

		case frame.Event == "pusher:ping":
			_ = writeJSON(map[string]any{"event": "pusher:pong", "data": map[string]any{}})

		case frame.Event == "App\\Events\\ChatMessage":
			msg, ok := parseKickMessage(frame.Channel, frame.Data, rooms)
			if ok {
				hub.Publish(msg)
			}
		}
	}
}

func parseKickMessage(channel, dataStr string, rooms map[string]string) (Message, bool) {
	var payload struct {
		ID      string `json:"id"`
		Content string `json:"content"`
		Sender  struct {
			Username string `json:"username"`
			Identity struct {
				Color string `json:"color"`
			} `json:"identity"`
		} `json:"sender"`
	}
	if err := json.Unmarshal([]byte(dataStr), &payload); err != nil {
		return Message{}, false
	}

	// channel looks like "chatrooms.{id}.v2" -> map id back to slug label.
	label := channel
	parts := strings.Split(channel, ".")
	if len(parts) >= 2 {
		if slug, ok := rooms[parts[1]]; ok {
			label = slug
		}
	}

	id := payload.ID
	if id == "" {
		id = fmt.Sprintf("ki-%s-%d-%s", label, nowMs(), randID())
	}

	return Message{
		ID:      id,
		Source:  "kick",
		Channel: label,
		Author:  payload.Sender.Username,
		Text:    stripNewlines(payload.Content),
		Color:   payload.Sender.Identity.Color,
		TS:      nowMs(),
	}, true
}

File: 02-websocket-hub/server/connectors/kick.js
'use strict';

const WebSocket = require('ws');
const { makeMessage } = require('../schema');

const PUSHER_URL =
  'wss://ws-us2.pusher.com/app/32cbd69e4b950bf97679?protocol=7&client=js&version=8.4.0&flash=false';

const BROWSER_UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36';

// Resolve a chatroom id for a slug. Kick sits behind Cloudflare, so send a browser UA.
async function resolveChatroomId(slug, log) {
  try {
    const res = await fetch(`https://kick.com/api/v2/channels/${encodeURIComponent(slug)}`, {
      headers: { 'User-Agent': BROWSER_UA, Accept: 'application/json' },
    });
    if (!res.ok) {
      log(`kick: lookup for "${slug}" returned ${res.status} (Cloudflare?). Use KICK_CHATROOM_IDS to override.`);
      return null;
    }
    const json = await res.json();
    const id = json && json.chatroom && json.chatroom.id;
    if (!id) {
      log(`kick: no chatroom id in response for "${slug}"`);
      return null;
    }
    return { id, slug };
  } catch (err) {
    log(`kick: lookup failed for "${slug}": ${err.message}`);
    return null;
  }
}

// channels = array of slugs. chatroomIdOverrides = array aligned to channels (optional).
function startKick(channels, chatroomIdOverrides, emit, log) {
  const slugs = (channels || []).map((c) => c.trim()).filter(Boolean);
  const overrides = chatroomIdOverrides || [];
  if (slugs.length === 0) return () => {};

  let ws = null;
  let closed = false;
  let attempt = 0;
  let reconnectTimer = null;
  // rooms: array of { id, slug }
  let rooms = [];

  function backoff() {
    attempt += 1;
    return Math.min(30000, 1000 * 2 ** Math.min(attempt, 5));
  }

  async function resolveRooms() {
    const out = [];
    for (let i = 0; i < slugs.length; i++) {
      const slug = slugs[i];
      const override = overrides[i] && String(overrides[i]).trim();
      if (override) {
        out.push({ id: override, slug });
        continue;
      }
      const r = await resolveChatroomId(slug, log);
      if (r) out.push(r);
    }
    return out;
  }

  function connect() {
    if (closed) return;
    ws = new WebSocket(PUSHER_URL);

    ws.on('open', () => {
      attempt = 0;
      log(`kick: pusher connected (${rooms.length} room(s))`);
    });

    ws.on('message', (buf) => {
      let frame;
      try {
        frame = JSON.parse(buf.toString('utf8'));
      } catch (_) {
        return;
      }

      if (frame.event === 'pusher:connection_established') {
        for (const room of rooms) {
          ws.send(
            JSON.stringify({
              event: 'pusher:subscribe',
              data: { auth: '', channel: `chatrooms.${room.id}.v2` },
            })
          );
        }
        log(`kick: subscribed to ${rooms.map((r) => r.slug).join(', ')}`);
        return;
      }

      if (frame.event === 'pusher:ping') {
        ws.send(JSON.stringify({ event: 'pusher:pong', data: {} }));
        return;
      }

      if (frame.event === 'App\\Events\\ChatMessage') {
        let data;
        try {
          data = typeof frame.data === 'string' ? JSON.parse(frame.data) : frame.data;
        } catch (_) {
          return;
        }
        // map channel back to slug via the chatrooms.{id}.v2 pusher channel name
        let slug = '';
        const m = /chatrooms\.(\d+)\.v2/.exec(frame.channel || '');
        if (m) {
          const found = rooms.find((r) => String(r.id) === m[1]);
          slug = found ? found.slug : m[1];
        }
        const sender = data.sender || {};
        emit(
          makeMessage({
            id: data.id || `kick-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`,
            source: 'kick',
            channel: slug,
            author: sender.username,
            text: data.content,
            color: (sender.identity && sender.identity.color) || '',
            ts: data.created_at ? Date.parse(data.created_at) || Date.now() : Date.now(),
          })
        );
      }
    });

    ws.on('close', () => {
      if (closed) return;
      const wait = backoff();
      log(`kick: pusher closed, reconnecting in ${wait}ms`);
      reconnectTimer = setTimeout(connect, wait);
    });

    ws.on('error', (err) => {
      log(`kick: error ${err.message}`);
      try { ws.close(); } catch (_) {}
    });
  }

  (async () => {
    rooms = await resolveRooms();
    if (rooms.length === 0) {
      log('kick: no resolvable chatrooms — connector idle (set KICK_CHATROOM_IDS to force).');
      return;
    }
    connect();
  })();

  return function stop() {
    closed = true;
    if (reconnectTimer) clearTimeout(reconnectTimer);
    try { ws && ws.close(); } catch (_) {}
  };
}

module.exports = { startKick };
